### ISSUE TITLE
Add type hints to `eth.tools.builder` modules

### ISSUE THREAD

**Issue by pipermerriam** [ethereum/py-evm#1366]
### What is wrong?

The `eth.tools.builder` module is not type hinted.  This makes using it in the `trinity` module not possible.

### How can it be fixed

Add type hints.

**Comment by Bhargavasomu** [comment]
Are the type hints needed for only the Functions or even the arguments to the function need to be type hinted?

**Comment by pipermerriam** [comment]
The whole module, functions, arguments, and anything else mypy complains about should be type hinted.

**Comment by Bhargavasomu** [comment]
Is there any easy way to find the types of the variables, or is it just by looking at the variables, we have to decide?

**Comment by cburgdorf** [comment]
>Is there any easy way to find the types of the variables, or is it just by looking at the variables, we have to decide?

There are tools such as [MonkeyType](https://github.com/Instagram/MonkeyType) but I'd recommend to just add them by hand. However, sometimes using `reveal_type` and `reveal_locals` can help to unveil what mypy *thinks* some types are. Just drop in `reveal_type(some_variable)` or `reveal_locals` and then run mypy.

For instance:

```python
class Foo:
    pass

class Bar:
    pass

x = 1
y = "Foo"

foo = Foo()
bar = Bar()

reveal_locals()
```

Running mypy on this gives me:

```
reveal_locals.py:13: error: Revealed local types are:
reveal_locals.py:13: error: bar: reveal_locals.Bar
reveal_locals.py:13: error: foo: reveal_locals.Foo
reveal_locals.py:13: error: x: builtins.int
reveal_locals.py:13: error: y: builtins.str
```

**Comment by Bhargavasomu** [comment]
@cburgdorf using reveal_locals or reveal_type is only givng the type as Any for all the definitions, since none of them are hinted anywhere. So I don't think that it is useful for this purpose.

**Comment by cburgdorf** [comment]
@Bhargavasomu This was more a general advice on how to work your way through with mypy. It obviously only works in scenarios where mypy can already infer a lot of the types. E.g. in the (trivial) example that I gave there are no type hints either, yet mypy inferred all types correctly.

**Comment by Bhargavasomu** [comment]
@cburgdorf I added the hints for most of the functions, but I am still getting the same error when I do this.
`mypy --follow-imports=silent --ignore-missing-imports --no-strict-optional --check-untyped-defs --disallow-incomplete-defs --disallow-untyped-defs --disallow-any-generics scripts/benchmark`

The error is as follows
`scripts/benchmark/utils/chain_plumbing.py:80 error: Cannot call function of unknown type`
The error is still occurring even though I added the type hints. So I ran the following command
`mypy --follow-imports=silent --ignore-missing-imports chain_plumbing.py`
I got no error while doing the above command, but got the error only when executing the whole benchmarks altogether. 
Is there something that we could infer from this?
